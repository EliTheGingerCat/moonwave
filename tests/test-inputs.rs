use std::path::Path;
use std::process::{Command, Stdio};

#[test]
fn everything_sandwich() -> anyhow::Result<()> {
    run_moonwave("passing/everything_sandwich.lua", 0)
}

#[test]
fn class_with_function() -> anyhow::Result<()> {
    run_moonwave("passing/class_with_function.lua", 0)
}

#[test]
fn fabric() -> anyhow::Result<()> {
    run_moonwave("passing/fabric.lua", 0)
}

#[test]
fn failing_function() -> anyhow::Result<()> {
    run_moonwave("failing/function.lua", 1)
}

#[test]
fn failing_function_no_within() -> anyhow::Result<()> {
    run_moonwave("failing/function_no_within.lua", 1)
}

fn run_moonwave(file_name: &str, expected_status: i32) -> anyhow::Result<()> {
    let path = Path::new("test-input").join(file_name);

    let child = Command::new(env!("CARGO_BIN_EXE_moonwave"))
        .arg("extract")
        .arg(path)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .env("NO_COLOR", "1")
        .spawn()?;

    let output = child.wait_with_output()?;
    let stdout = String::from_utf8(output.stdout)?;
    let stderr = String::from_utf8(output.stderr)?;

    let stdout_name = format!("{}-stdout", file_name);
    let stderr_name = format!("{}-stderr", file_name);

    insta::assert_snapshot!(stdout_name, stdout);
    insta::assert_snapshot!(stderr_name, stderr);

    assert_eq!(output.status.code(), Some(expected_status));

    Ok(())
}
