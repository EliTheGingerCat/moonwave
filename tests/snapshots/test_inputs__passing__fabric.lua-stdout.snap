---
source: tests/test-inputs.rs
expression: stdout
---
[
  {
    "functions": [
      {
        "name": "new",
        "desc": "Creates a new instance of Fabric.",
        "params": [
          {
            "name": "namespace",
            "desc": "A unique namespace to distinguish from other instances of Fabric for network calls.",
            "lua_type": "string"
          }
        ],
        "returns": [
          {
            "desc": "",
            "lua_type": "Fabric"
          }
        ],
        "tags": [],
        "errors": [],
        "function_type": "static",
        "realm": [],
        "private": false,
        "unreleased": false,
        "yields": false,
        "ignore": false
      },
      {
        "name": "registerUnit",
        "desc": "Registers a unit. This function should be called before attempting to get or create the unit.",
        "params": [
          {
            "name": "unitDefinition",
            "desc": "The definition of the unit",
            "lua_type": "UnitDefinition"
          }
        ],
        "returns": [
          {
            "desc": "The passed unit definition",
            "lua_type": "UnitDefinition"
          }
        ],
        "tags": [],
        "errors": [],
        "function_type": "method",
        "realm": [],
        "private": false,
        "unreleased": false,
        "yields": false,
        "ignore": false
      },
      {
        "name": "registerUnitsIn",
        "desc": "Registers all units that are immmediate children of a container.\nSkips any test scripts (i.e. name of form `*.spec`) in the container.",
        "params": [
          {
            "name": "container",
            "desc": "The container",
            "lua_type": "Instance"
          }
        ],
        "returns": [
          {
            "desc": "",
            "lua_type": "nil"
          }
        ],
        "tags": [],
        "errors": [],
        "function_type": "method",
        "realm": [],
        "private": false,
        "unreleased": false,
        "yields": false,
        "ignore": false
      },
      {
        "name": "getUnitByRef",
        "desc": "Returns the unit associated with a unit resolvable that is attached to a ref,\nor nil if it doesn't exist.",
        "params": [
          {
            "name": "unitResolvable",
            "desc": "The unit to retrieve",
            "lua_type": "UnitResolvable"
          },
          {
            "name": "ref",
            "desc": "The ref to retrieve the unit from",
            "lua_type": "Ref"
          }
        ],
        "returns": [
          {
            "desc": "The attached unit",
            "lua_type": "Unit?"
          }
        ],
        "tags": [],
        "errors": [],
        "function_type": "method",
        "realm": [],
        "private": false,
        "unreleased": false,
        "yields": false,
        "ignore": false
      },
      {
        "name": "getOrCreateUnitByRef",
        "desc": "Returns the unit associated with a unit resolvable that is attached to ref.\nIf it does not exist, then creates and attaches the unit to ref and returns it.",
        "params": [
          {
            "name": "unitResolvable",
            "desc": "The unit to retrieve",
            "lua_type": "UnitResolvable"
          },
          {
            "name": "ref",
            "desc": "The ref to retrieve the attached unit from",
            "lua_type": "Ref"
          }
        ],
        "returns": [
          {
            "desc": "The attached unit",
            "lua_type": "Unit"
          }
        ],
        "tags": [],
        "errors": [],
        "function_type": "method",
        "realm": [],
        "private": false,
        "unreleased": false,
        "yields": false,
        "ignore": false
      },
      {
        "name": "removeAllUnitsWithRef",
        "desc": "Removes all units attached to the passed ref.",
        "params": [
          {
            "name": "ref",
            "desc": "The ref to remove all units from",
            "lua_type": "Ref"
          }
        ],
        "returns": [
          {
            "desc": "",
            "lua_type": "nil"
          }
        ],
        "tags": [],
        "errors": [],
        "function_type": "method",
        "realm": [],
        "private": false,
        "unreleased": false,
        "yields": false,
        "ignore": false
      },
      {
        "name": "fire",
        "desc": "Fires a fabric event.",
        "params": [
          {
            "name": "eventName",
            "desc": "The event name to fire",
            "lua_type": "string"
          },
          {
            "name": "...",
            "desc": "The arguments to fire the event with.",
            "lua_type": "any"
          }
        ],
        "returns": [
          {
            "desc": "",
            "lua_type": "nil"
          }
        ],
        "tags": [],
        "errors": [],
        "function_type": "method",
        "realm": [],
        "private": false,
        "unreleased": false,
        "yields": false,
        "ignore": false
      },
      {
        "name": "on",
        "desc": "Listens to a fabric event.",
        "params": [
          {
            "name": "eventName",
            "desc": "The event name to listen to",
            "lua_type": "string"
          },
          {
            "name": "callback",
            "desc": "The callback fired",
            "lua_type": "function"
          }
        ],
        "returns": [
          {
            "desc": "",
            "lua_type": "nil"
          }
        ],
        "tags": [],
        "errors": [],
        "function_type": "method",
        "realm": [],
        "private": false,
        "unreleased": false,
        "yields": false,
        "ignore": false
      },
      {
        "name": "debug",
        "desc": "Logs a debug message. Set fabric.DEBUG = true to enable.",
        "params": [
          {
            "name": "...",
            "desc": "The debug information to log",
            "lua_type": "any"
          }
        ],
        "returns": [
          {
            "desc": "",
            "lua_type": "nil"
          }
        ],
        "tags": [],
        "errors": [],
        "function_type": "method",
        "realm": [],
        "private": false,
        "unreleased": false,
        "yields": false,
        "ignore": false
      }
    ],
    "properties": [
      {
        "name": "reducers",
        "desc": "This is a property that is set to the reducers.",
        "tags": [],
        "realm": [],
        "private": false,
        "unreleased": false,
        "readonly": true,
        "ignore": false
      }
    ],
    "types": [
      {
        "name": "stringOrNil",
        "desc": "A string or nil.",
        "tags": [],
        "private": false,
        "ignore": false
      }
    ],
    "name": "Fabric",
    "desc": ".reducers Reducers -- This is a property that is set to the reducers.",
    "tags": [],
    "realm": [],
    "private": false,
    "unreleased": false,
    "ignore": false
  }
]

